src/self-managed-debug/src/main.rs

```diff
@@ -0,0 +1,348 @@
+    };
+
+    let kubeconfig = Config::from_kubeconfig(&kubeconfig_options).await?;
+
+    let client = Client::try_from(kubeconfig)?;
+
+    Ok(client)
+}
+
+/// Write k8s pod logs to a file per pod.
+async fn dump_k8s_pod_logs(
+    context: &Context,
+    client: Client,
+    namespace: &String,
+) -> Result<(), anyhow::Error> {
+    let file_path = format_resource_path(context.start_time, namespace, &K8sResourceType::Log);
+    create_dir_all(&file_path)?;
+
+    let pods: Api<Pod> = Api::<Pod>::namespaced(client.clone(), namespace);
+
+    let pod_list = pods.list(&ListParams::default()).await?;
+
+    for pod in &pod_list.items {
+        let pod_name = pod.metadata.name.clone().unwrap_or_default();
+        let previous_logs_file_name = format!("{}/{}.previous.log", file_path, pod_name);
+        let current_logs_file_name = format!("{}/{}.current.log", file_path, pod_name);
+
+        if let Err(e) = async {
+            let previous_logs = pods
+                .logs(
+                    &pod_name,
+                    &LogParams {
+                        previous: true,
+                        timestamps: true,
+                        ..Default::default()
+                    },
+                )
+                .await?;
+
+            if previous_logs.is_empty() {
+                println!("No previous logs found for pod {}", pod_name);
+                return Ok(());
+            }
+
+            let mut file = File::create(&previous_logs_file_name)?;
+
+            file.write_all(previous_logs.as_bytes())?;
+
+            println!("Exported {}", &previous_logs_file_name);
+            Ok::<(), anyhow::Error>(())
+        }
+        .await
+        {
+            if let Some(kube::Error::Api(e)) = e.downcast_ref::<kube::Error>() {
+                if e.code == 400 {
+                    eprintln!("No previous logs available for pod {}", pod_name);
+                } else {
+                    eprintln!("Failed to export {}: {}", &previous_logs_file_name, e);
+                }
+            } else {
+                eprintln!("Failed to export {}: {}", &previous_logs_file_name, e);
+            }
+        }
+
+        if let Err(e) = async {
+            let logs = pods
+                .logs(
+                    &pod_name,
+                    &LogParams {
+                        timestamps: true,
+                        ..Default::default()
+                    },
+                )
+                .await?;
+
+            if logs.is_empty() {
+                println!("No logs found for pod {}", pod_name);
+                return Ok(());
+            }
+            let mut file = File::create(&current_logs_file_name)?;
+
+            file.write_all(logs.as_bytes())?;
+
+            println!("Exported {}", &current_logs_file_name);
+            Ok::<(), anyhow::Error>(())
+        }
+        .await
+        {
+            eprintln!("Failed to export {}: {}", &current_logs_file_name, e);
+        }
```

Comment:
There is a lot of duplication here. Maybe pull this all out into a separate function that is called for both the previous and current logs?